cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -454,12 +531,49 @@ public void saveReport(HttpServiceRequest request, HttpServiceResponder responde
                            StandardCharsets.UTF_8);
     }
 
+    /**
+     * Use reportId and userName to construct the {@link ReportIdentifier}
+     * If instead shareId is provided, we decode the userName and reportId from shareId and construct the identifier
+     * and return that.
+     * @param reportId - id of the report, can be null if shareId is provided
+     * @param headers - request headers, used to retrieve the name of the owner who generated the report,
+     *                  if reportId is not null
+     * @param shareId - shareId if the report is a shared report, can be null if reportId and userName is provided
+     * @return {@link ReportIdentifier}
+     * @throws IllegalArgumentException if none of the reportId and shareId is provided or if both are provided
+     * @throws UnsupportedEncodingException if there are any error while decoding the shareId
+     */
+    private ReportIdentifier validateAndGetReportIdentifier(
```

Comment:
can combine this method with `getLocationFromReportIdentifier`, looks like the ReportIdentifier returned by this method is never used at other places except `getLocationFromReportIdentifier`

Reply:
No, by having separate methods, it gives better separation of context, method does one thing than multiple things and able to respond with better error message when there is an `IOException`.